// @mnemopay/sdk/dist/commerce/checkout/executor.js
"use strict";
/**
 * Checkout Executor — orchestrates the full purchase flow
 *
 * Coordinates between:
 *   - CommerceProvider (finds the product)
 *   - CheckoutStrategy (completes the purchase on the merchant site)
 *   - BuyerProfile (shipping + payment info)
 *   - MnemoPay (escrow, approval, settlement)
 *
 * Usage:
 *   const executor = new CheckoutExecutor({ profile, screenshotDir });
 *   const result = await executor.checkout(productUrl);
 *
 * The executor does NOT require a live browser — it creates one on demand
 * via dynamic import of playwright. If playwright is not installed,
 * executePurchase() returns checkout_required with the URL.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckoutExecutor = void 0;
const profile_js_1 = require("./profile.js");
const shopify_js_1 = require("./strategies/shopify.js");
const generic_js_1 = require("./strategies/generic.js");
// ── Executor ───────────────────────────────────────────────────────────────
class CheckoutExecutor {
    profile;
    screenshotDir;
    strategies;
    timeout;
    headless;
    userAgent;
    constructor(config) {
        // Validate profile
        const errors = (0, profile_js_1.validateProfile)(config.profile);
        if (errors.length > 0) {
            throw new Error(`Invalid buyer profile: ${errors.join(", ")}`);
        }
        this.profile = config.profile;
        this.screenshotDir = config.screenshotDir;
        this.timeout = config.timeout ?? 30_000;
        this.headless = config.headless ?? true;
        this.userAgent = config.userAgent ??
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36";
        // Strategy chain: custom strategies first, then Shopify, then generic fallback
        this.strategies = [
            ...(config.strategies ?? []),
            new shopify_js_1.ShopifyCheckoutStrategy(),
            new generic_js_1.GenericCheckoutStrategy(),
        ];
    }
    /**
     * Execute a full checkout for the given product URL.
     *
     * Flow:
     *   1. Launch browser
     *   2. Select best checkout strategy for the URL
     *   3. Execute checkout steps
     *   4. Return result with order ID, screenshots, timing
     */
    async checkout(productUrl) {
        // Input validation
        if (!productUrl || typeof productUrl !== "string") {
            return { success: false, failureReason: "Invalid product URL", steps: [], elapsedMs: 0 };
        }
        // URL sanitization — only allow http/https
        let parsedUrl;
        try {
            parsedUrl = new URL(productUrl);
            if (!["http:", "https:"].includes(parsedUrl.protocol)) {
                return { success: false, failureReason: "Only HTTP/HTTPS URLs are supported", steps: [], elapsedMs: 0 };
            }
        }
        catch {
            return { success: false, failureReason: "Malformed URL", steps: [], elapsedMs: 0 };
        }
        // Find matching strategy
        const strategy = this.strategies.find(s => s.canHandle(productUrl));
        if (!strategy) {
            return { success: false, failureReason: "No checkout strategy can handle this URL", steps: [], elapsedMs: 0 };
        }
        // Dynamic import — playwright is optional
        let playwright;
        try {
            playwright = await import("playwright");
        }
        catch {
            return { success: false, failureReason: "Playwright not installed. Run: npm install playwright", steps: [], elapsedMs: 0 };
        }
        const browser = await playwright.chromium.launch({
            headless: this.headless,
        });
        const screenshots = [];
        const logs = [];
        try {
            const context = await browser.newContext({
                userAgent: this.userAgent,
                viewport: { width: 1280, height: 800 },
                locale: "en-US",
            });
            // Block unnecessary resources for speed
            await context.route("**/*.{png,jpg,jpeg,gif,svg,mp4,webm,woff,woff2,ttf}", (route) => route.abort());
            const page = await context.newPage();
            page.setDefaultTimeout(this.timeout);
            const ctx = {
                page,
                productUrl,
                log: (msg) => {
                    logs.push(`[${strategy.name}] ${msg}`);
                },
                screenshot: async (name) => {
                    if (this.screenshotDir) {
                        const path = `${this.screenshotDir}/${name}.png`;
                        await page.screenshot({ path, fullPage: false }).catch(() => { });
                        screenshots.push(path);
                    }
                },
            };
            // Global timeout: 3x the page timeout to cover entire checkout flow
            const globalTimeout = this.timeout * 3;
            const result = await Promise.race([
                strategy.execute(ctx, this.profile),
                new Promise((_, reject) => setTimeout(() => reject(new Error(`Checkout timed out after ${globalTimeout}ms`)), globalTimeout)),
            ]);
            result.screenshots = screenshots;
            // Log summary
            if (result.success) {
                logs.push(`Checkout completed: order ${result.orderId} in ${result.elapsedMs}ms`);
            }
            else {
                logs.push(`Checkout failed: ${result.failureReason}`);
            }
            return result;
        }
        finally {
            await browser.close();
        }
    }
    /**
     * Test connectivity to a merchant site without purchasing.
     * Returns the detected strategy name and page title.
     */
    async probe(url) {
        let playwright;
        try {
            playwright = await import("playwright");
        }
        catch {
            return { strategy: "none", title: "", canCheckout: false };
        }
        const browser = await playwright.chromium.launch({ headless: true });
        try {
            const page = await browser.newPage();
            await page.goto(url, { waitUntil: "domcontentloaded", timeout: 15_000 });
            const title = await page.title();
            const strategy = this.strategies.find(s => s.canHandle(url));
            return {
                strategy: strategy?.name ?? "none",
                title,
                canCheckout: !!strategy,
            };
        }
        finally {
            await browser.close();
        }
    }
}
exports.CheckoutExecutor = CheckoutExecutor;
