// @mnemopay/sdk/dist/rails/index.js
"use strict";
/**
 * Payment Rail Abstraction
 *
 * Pluggable payment backends for MnemoPay. The default MockRail keeps
 * the existing in-memory ledger behavior. Real rails (Stripe, Lightning)
 * connect to actual payment processors.
 *
 * Usage:
 *   const agent = MnemoPay.quick("id", { paymentRail: new StripeRail("sk_test_...") });
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightningRail = exports.NIGERIAN_BANKS = exports.PaystackRail = exports.StripeRail = exports.MockRail = void 0;
// ─── Mock Rail (default — existing in-memory behavior) ──────────────────────
class MockRail {
    name = "mock";
    counter = 0;
    async createHold(amount, reason, agentId, _opts) {
        return {
            externalId: `mock_hold_${++this.counter}`,
            status: "held",
        };
    }
    async capturePayment(externalId, amount) {
        return {
            externalId,
            status: "captured",
            receiptId: `mock_receipt_${this.counter}`,
        };
    }
    async reversePayment(externalId, amount) {
        return {
            externalId,
            status: "reversed",
        };
    }
}
exports.MockRail = MockRail;
// ─── Stripe Rail ────────────────────────────────────────────────────────────
// Requires: npm install stripe (peer dependency)
// Uses PaymentIntents with manual capture for true escrow.
class StripeRail {
    name = "stripe";
    stripe;
    currency;
    inFlightCaptures = new Map();
    /**
     * @param secretKey — Stripe secret key (sk_test_... or sk_live_...)
     * @param currency — ISO currency code (default: "usd")
     */
    constructor(secretKey, currency = "usd") {
        this.currency = currency;
        try {
            const Stripe = require("stripe");
            this.stripe = new Stripe(secretKey);
        }
        catch {
            throw new Error("Stripe package not installed. Run: npm install stripe");
        }
    }
    /**
     * Build a StripeRail from an already-constructed Stripe client.
     *
     * Useful for tests (inject a mock) and for apps that want to share a
     * single Stripe client instance across multiple rails / services.
     */
    static fromClient(client, currency = "usd") {
        // Bypass the require() path by stamping the field on a bare instance.
        const rail = Object.create(StripeRail.prototype);
        rail.name = "stripe";
        rail.stripe = client;
        rail.currency = currency;
        rail.inFlightCaptures = new Map();
        return rail;
    }
    async createHold(amount, reason, agentId, opts) {
        const params = {
            amount: Math.round(amount * 100), // Stripe uses cents
            currency: this.currency,
            capture_method: "manual", // Hold funds, don't capture yet
            metadata: {
                agentId,
                reason: reason.slice(0, 500),
                source: "mnemopay",
                ...(opts?.metadata ?? {}),
            },
        };
        // Target a specific customer + saved payment method when provided.
        // This is the path that actually moves money without a browser handoff.
        if (opts?.customerId)
            params.customer = opts.customerId;
        if (opts?.paymentMethodId) {
            params.payment_method = opts.paymentMethodId;
            // Only confirm automatically when we have a PM to confirm with.
            // Omitting confirm lets the caller attach a PM client-side (Stripe.js)
            // and confirm from the browser — the pre-v1.0 flow.
            params.confirm = true;
            if (opts.offSession)
                params.off_session = true;
        }
        const idempotencyKey = opts?.metadata?.idempotencyKey ||
            opts?.metadata?.requestId;
        const intent = await this.stripe.paymentIntents.create(params, idempotencyKey ? { idempotencyKey } : {});
        return {
            externalId: intent.id,
            status: intent.status,
        };
    }
    async capturePayment(externalId, amount) {
        const existing = this.inFlightCaptures.get(externalId);
        if (existing)
            return existing;
        const promise = (async () => {
            const intent = await this.stripe.paymentIntents.capture(externalId, {
                amount_to_capture: Math.round(amount * 100),
            }, {
                idempotencyKey: `cap_${externalId}`,
            });
            return {
                externalId: intent.id,
                status: intent.status,
                receiptId: intent.latest_charge,
            };
        })();
        this.inFlightCaptures.set(externalId, promise);
        try {
            return await promise;
        }
        finally {
            this.inFlightCaptures.delete(externalId);
        }
    }
    async reversePayment(externalId, _amount) {
        const intent = await this.stripe.paymentIntents.cancel(externalId);
        return {
            externalId: intent.id,
            status: intent.status === "canceled" ? "reversed" : intent.status,
        };
    }
    // ── Customer onboarding helpers ─────────────────────────────────────────
    // Thin wrappers around Stripe's customer + SetupIntent APIs. These let
    // apps collect and save a card without pulling in the full Stripe SDK
    // from every service that only needs MnemoPay.
    /**
     * Create a Stripe customer. Returns the customer id (cus_...).
     * Persist this against your internal user/agent record.
     */
    async createCustomer(email, name, metadata) {
        if (!email || typeof email !== "string") {
            throw new Error("email is required");
        }
        const customer = await this.stripe.customers.create({
            email,
            ...(name ? { name } : {}),
            ...(metadata ? { metadata } : {}),
        });
        return { customerId: customer.id };
    }
    /**
     * Create a SetupIntent for collecting a card via Stripe.js (off-session
     * charges later). Return the client_secret to the frontend and confirm
     * it with Stripe.js — no card data ever touches your servers.
     */
    async createSetupIntent(customerId) {
        if (!customerId || typeof customerId !== "string") {
            throw new Error("customerId is required");
        }
        const si = await this.stripe.setupIntents.create({
            customer: customerId,
            usage: "off_session",
        });
        return { setupIntentId: si.id, clientSecret: si.client_secret };
    }
}
exports.StripeRail = StripeRail;
// Re-export Paystack rail
var paystack_js_1 = require("./paystack.js");
Object.defineProperty(exports, "PaystackRail", { enumerable: true, get: function () { return paystack_js_1.PaystackRail; } });
Object.defineProperty(exports, "NIGERIAN_BANKS", { enumerable: true, get: function () { return paystack_js_1.NIGERIAN_BANKS; } });
// ─── Lightning Rail (L402) ──────────────────────────────────────────────────
// Requires a running LND node. Uses HODL invoices for escrow.
class LightningRail {
    name = "lightning";
    baseUrl;
    macaroon;
    btcPriceUsd;
    /**
     * Check whether a hostname resolves to a private, reserved, or otherwise
     * dangerous network address. Covers IPv6-mapped IPv4, numeric/hex/octal
     * IP bypass tricks, cloud metadata endpoints, and RFC1918 ranges.
     */
    static isPrivateOrReserved(hostname) {
        const h = hostname.toLowerCase().replace(/\[|\]/g, "");
        // Direct matches
        const blocked = [
            "localhost", "0.0.0.0", "[::]", "::1", "::ffff:127.0.0.1",
            "metadata.google.internal", "169.254.169.254",
        ];
        if (blocked.some(b => h === b || h.includes(b)))
            return true;
        // IPv4 numeric/hex/octal bypass
        // Parse as number — if it resolves to a private range, block it
        const asNum = Number(h);
        if (!isNaN(asNum) && asNum >= 0)
            return true; // Block ALL numeric IPs
        // Hex IPs
        if (/^0x[0-9a-f]+$/i.test(h))
            return true;
        // Octal IPs (starts with 0)
        if (/^0\d+(\.\d+){0,3}$/.test(h))
            return true;
        // IPv6-mapped IPv4
        if (/::ffff:/i.test(h))
            return true;
        // IPv6 loopback
        if (h === "::1" || h === "[::1]")
            return true;
        // Suffix-based blocks
        if (h.endsWith(".internal") || h.endsWith(".local"))
            return true;
        // RFC1918 ranges and other reserved IPv4
        const parts = h.split(".").map(Number);
        if (parts.length === 4 && parts.every(p => !isNaN(p) && p >= 0 && p <= 255)) {
            if (parts[0] === 127)
                return true;
            if (parts[0] === 10)
                return true;
            if (parts[0] === 172 && parts[1] >= 16 && parts[1] <= 31)
                return true;
            if (parts[0] === 192 && parts[1] === 168)
                return true;
            if (parts[0] === 169 && parts[1] === 254)
                return true;
            if (parts[0] === 0)
                return true;
        }
        return false;
    }
    /**
     * @param lndRestUrl — LND REST API URL (e.g. "https://localhost:8080")
     * @param macaroon — Admin macaroon hex string
     * @param btcPriceUsd — BTC price in USD for conversion (updated externally)
     */
    constructor(lndRestUrl, macaroon, btcPriceUsd = 60000) {
        // Validate URL to prevent SSRF
        let parsed;
        try {
            parsed = new URL(lndRestUrl);
        }
        catch {
            throw new Error("Invalid LND REST URL");
        }
        if (parsed.protocol !== "https:" && parsed.protocol !== "http:") {
            throw new Error("LND URL must use http or https protocol");
        }
        // SSRF protection: block private/internal network targets
        if (LightningRail.isPrivateOrReserved(parsed.hostname)) {
            throw new Error("LND URL must not target private/internal network addresses");
        }
        this.baseUrl = `${parsed.protocol}//${parsed.host}${parsed.pathname.replace(/\/$/, "")}`;
        this.macaroon = macaroon;
        this.btcPriceUsd = btcPriceUsd;
    }
    /** Update BTC price for USD→sats conversion */
    setBtcPrice(usd) {
        this.btcPriceUsd = usd;
    }
    usdToSats(usd) {
        return Math.round((usd / this.btcPriceUsd) * 100_000_000);
    }
    async lndRequest(path, method, body) {
        const res = await fetch(`${this.baseUrl}${path}`, {
            method,
            headers: {
                "Grpc-Metadata-macaroon": this.macaroon,
                "Content-Type": "application/json",
            },
            body: body ? JSON.stringify(body) : undefined,
        });
        if (!res.ok) {
            // Sanitize error — don't leak raw LND response body
            throw new Error(`LND API error (${res.status})`);
        }
        return res.json();
    }
    async createHold(amount, reason, agentId, _opts) {
        const sats = this.usdToSats(amount);
        // Create a standard invoice (HODL invoices need invoicesrpc)
        const invoice = await this.lndRequest("/v1/invoices", "POST", {
            value: sats.toString(),
            memo: `mnemopay:${agentId}:${reason.slice(0, 100)}`,
            expiry: "3600", // 1 hour
        });
        return {
            externalId: invoice.r_hash, // Base64 payment hash
            status: "invoice_created",
            receiptId: invoice.payment_request, // Lightning invoice string
        };
    }
    async capturePayment(externalId, _amount) {
        // externalId is the r_hash (base64 payment hash) from createHold.
        // LND REST requires the r_hash as URL-safe base64 in the path.
        const rHashUrlSafe = externalId
            .replace(/\+/g, "-")
            .replace(/\//g, "_")
            .replace(/=+$/, "");
        const invoice = await this.lndRequest(`/v1/invoice/${rHashUrlSafe}`, "GET");
        return {
            externalId,
            status: invoice.settled ? "captured" : "pending",
            receiptId: invoice.payment_request,
        };
    }
    async reversePayment(externalId, _amount) {
        // Lightning invoices can't be reversed once paid
        // Cancel unpaid invoice by letting it expire
        return {
            externalId,
            status: "expired",
        };
    }
}
exports.LightningRail = LightningRail;
