// @mnemopay/sdk/dist/client.js
"use strict";
/**
 * MnemoPay Universal Client
 *
 * Thin client that works everywhere: browser, React Native, Node.js, Deno.
 * Zero dependencies. Talks to a MnemoPay server via REST API.
 *
 * Usage:
 *   import { MnemoPayClient } from "@mnemopay/sdk/client";
 *
 *   const client = new MnemoPayClient("http://localhost:3200", "your-token");
 *   await client.remember("User prefers monthly billing");
 *   const memories = await client.recall("billing preferences");
 *   const tx = await client.charge(25, "Monthly access");
 *   await client.settle(tx.txId);
 *
 * Works from:
 *   - Browser (any modern browser with fetch)
 *   - React Native
 *   - Node.js 18+
 *   - Deno
 *   - Any HTTP client
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MnemoPayClient = void 0;
// ─── Client ─────────────────────────────────────────────────────────────────
class MnemoPayClient {
    baseUrl;
    token;
    timeoutMs;
    _fetch;
    constructor(baseUrl, token, config) {
        this.baseUrl = baseUrl.replace(/\/$/, "");
        this.token = token;
        this.timeoutMs = config?.timeoutMs ?? 30_000;
        this._fetch = config?.fetch ?? globalThis.fetch.bind(globalThis);
    }
    // ── Core HTTP ───────────────────────────────────────────────────────────
    async request(method, path, body) {
        const controller = new AbortController();
        const timeout = setTimeout(() => controller.abort(), this.timeoutMs);
        const headers = { "Content-Type": "application/json" };
        if (this.token)
            headers["Authorization"] = `Bearer ${this.token}`;
        try {
            const res = await this._fetch(`${this.baseUrl}${path}`, {
                method,
                headers,
                body: body !== undefined ? JSON.stringify(body) : undefined,
                signal: controller.signal,
            });
            const json = await res.json();
            if (!res.ok || json.ok === false) {
                throw new Error(json.error ?? `Request failed (${res.status})`);
            }
            return json.result !== undefined ? json.result : json;
        }
        catch (err) {
            if (err.name === "AbortError") {
                throw new Error(`Request timed out after ${this.timeoutMs}ms`);
            }
            throw err;
        }
        finally {
            clearTimeout(timeout);
        }
    }
    post(path, body) {
        return this.request("POST", path, body);
    }
    get(path) {
        return this.request("GET", path);
    }
    // ── Memory ──────────────────────────────────────────────────────────────
    /** Store a memory */
    async remember(content, options) {
        return this.post("/api/remember", { content, ...options });
    }
    /** Recall memories, optionally filtered by semantic query */
    async recall(queryOrLimit, limit) {
        if (typeof queryOrLimit === "number") {
            return this.post("/api/recall", { limit: queryOrLimit });
        }
        return this.post("/api/recall", { query: queryOrLimit, limit });
    }
    /** Delete a memory by ID */
    async forget(id) {
        return this.post("/api/forget", { id });
    }
    /** Boost a memory's importance */
    async reinforce(id, boost) {
        return this.post("/api/reinforce", { id, boost });
    }
    /** Prune stale memories */
    async consolidate() {
        return this.post("/api/consolidate", {});
    }
    // ── Payments ────────────────────────────────────────────────────────────
    /** Create an escrow charge */
    async charge(amount, reason) {
        return this.post("/api/charge", { amount, reason });
    }
    /** Settle a pending escrow */
    async settle(txId, counterpartyId) {
        return this.post("/api/settle", { txId, counterpartyId });
    }
    /** Refund a transaction */
    async refund(txId) {
        return this.post("/api/refund", { txId });
    }
    /** Check wallet balance and reputation */
    async balance() {
        return this.post("/api/balance", {});
    }
    /** Full agent profile */
    async profile() {
        return this.post("/api/profile", {});
    }
    /** Transaction history */
    async history(limit) {
        return this.post("/api/history", { limit });
    }
    /** Audit logs */
    async logs(limit) {
        return this.post("/api/logs", { limit });
    }
    /** Reputation report */
    async reputation() {
        return this.post("/api/reputation", {});
    }
    /** File a dispute */
    async dispute(txId, reason) {
        return this.post("/api/dispute", { txId, reason });
    }
    /** Fraud detection stats */
    async fraudStats() {
        return this.post("/api/fraud_stats", {});
    }
    // ── Commerce ────────────────────────────────────────────────────────────
    /** Set a shopping mandate (budget, categories, merchant restrictions) */
    async setMandate(mandate) {
        return this.request("POST", "/api/commerce/mandate", mandate);
    }
    /** Search for products within mandate constraints */
    async search(query, options) {
        return this.request("POST", "/api/commerce/search", { query, options });
    }
    /** Purchase a product (creates escrow, executes purchase) */
    async purchase(product, deliveryInstructions) {
        return this.request("POST", "/api/commerce/purchase", { product, deliveryInstructions });
    }
    /** Confirm delivery (settles escrow) */
    async confirmDelivery(orderId) {
        return this.request("POST", "/api/commerce/confirm", { orderId });
    }
    /** Cancel an order (refunds escrow) */
    async cancelOrder(orderId, reason) {
        return this.request("POST", "/api/commerce/cancel", { orderId, reason });
    }
    /** List orders, optionally filtered by status */
    async orders(status) {
        const qs = status ? `?status=${encodeURIComponent(status)}` : "";
        return this.request("GET", `/api/commerce/orders${qs}`);
    }
    // ── Utility ─────────────────────────────────────────────────────────────
    /** Check server health */
    async health() {
        return this.get("/health");
    }
    /** List available tools */
    async tools() {
        return this.get("/api/tools");
    }
}
exports.MnemoPayClient = MnemoPayClient;
exports.default = MnemoPayClient;
